app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js

```diff
@@ -234,91 +195,298 @@ async function addEthereumChainHandler(
+    }
   }
-  let networkConfigurationId;
-
-  const { id: approvalFlowId } = await startApprovalFlow();
+  return end();
+};
 
+const switchChainWithoutPermissions = async (
+  res,
+  end,
+  requestData,
+  networkConfigurationId,
+  approvalFlowId,
+  origin,
+  { setActiveNetwork, endApprovalFlow, requestUserApproval },
+) => {
   try {
     await requestUserApproval({
       origin,
-      type: ApprovalType.AddEthereumChain,
-      requestData: {
-        chainId: _chainId,
-        rpcPrefs: { blockExplorerUrl: firstValidBlockExplorerUrl },
-        chainName: _chainName,
-        rpcUrl: firstValidRPCUrl,
-        ticker,
-      },
+      type: ApprovalType.SwitchEthereumChain,
+      requestData,
     });
 
-    networkConfigurationId = await upsertNetworkConfiguration(
-      {
-        chainId: _chainId,
-        rpcPrefs: { blockExplorerUrl: firstValidBlockExplorerUrl },
-        nickname: _chainName,
-        rpcUrl: firstValidRPCUrl,
-        ticker,
-      },
-      { source: MetaMetricsNetworkEventSource.Dapp, referrer: origin },
-    );
-
-    // Once the network has been added, the requested is considered successful
+    await setActiveNetwork(networkConfigurationId);
     res.result = null;
   } catch (error) {
-    endApprovalFlow({ id: approvalFlowId });
-    return end(error);
+    // For the purposes of this method, it does not matter if the user
+    // declines to switch the selected network. However, other errors indicate
+    // that something is wrong.
+    if (error.code !== errorCodes.provider.userRejectedRequest) {
+      return end(error);
+    }
+  } finally {
+    if (approvalFlowId) {
+      endApprovalFlow({ id: approvalFlowId });
+    }
   }
+  return end();
+};
 
+const switchChain = async (
+  res,
+  end,
+  origin,
+  chainId,
+  requestData,
+  networkConfigurationId,
+  approvalFlowId,
+  {
+    getChainPermissionsFeatureFlag,
+    setActiveNetwork,
+    endApprovalFlow,
+    requestUserApproval,
+    getCaveat,
+    requestSwitchNetworkPermission,
+  },
+) => {
+  // If this network is already added but is not the currently selected network for the given origin
   // Ask the user to switch the network
-  try {
-    await requestUserApproval({
+  if (getChainPermissionsFeatureFlag()) {
+    await switchChainWithPermissions(
+      res,
+      end,
+      networkConfigurationId,
+      approvalFlowId,
+      chainId,
+      {
+        getCaveat,
+        requestSwitchNetworkPermission,
+        setActiveNetwork,
+        endApprovalFlow,
+      },
+    );
+  } else {
+    // TODO remove once CHAIN_PERMISSIONS feature flag is gone
+    await switchChainWithoutPermissions(
+      res,
+      end,
+      requestData,
+      networkConfigurationId,
+      approvalFlowId,
       origin,
-      type: ApprovalType.SwitchEthereumChain,
-      requestData: {
-        toNetworkConfiguration: {
+      { setActiveNetwork, endApprovalFlow, requestUserApproval },
+    );
+  }
+  return end();
+};
```

Comment:
DRY: This is what the heart of the `switchEthereumChain` implementation should look like, and it should be extracted into a separate module. We should be able to whittle away `switch-ethereum-chain.js` to almost nothing by creating the separate module with shared code that I suggest in this file.

Reply:
Did some general DRYing here: https://github.com/MetaMask/metamask-extension/pull/24415/commits/9d07ed684b9484b5f42a3baea873a39aa413be7e 

and here: https://github.com/MetaMask/metamask-extension/pull/24415/commits/46d4528175765cf4603941ac2913b75fdde1030f